File: problem_solving_with_DSA/stack/Python/stringRev.py
"""Reverse string using stack data structure"""

size = 10
stack = ['' for _ in range(size)]
top = -1

def push(ch: str) -> None:
    """Push ch into stack"""
    global top
    if top == size-1:
        return
    else:
        top += 1
        print(top)
        stack[top] = ch

def pop() -> int:
    """Pop off the top data in the stack"""
    global top
    if top == -1:
        return -1
    else:
        data = stack[top]
        top -= 1
        return data

def stringrev(s) :
    """Reverse string using stack data structure"""
    temp = ''
    for ch in s:
        push(ch)
        
    for _ in range(len(s)):
        temp += pop()
        
    return temp
    

if __name__ == '__main__':
    value = "Apple"
    
    print(value)
    print(stringrev(value))
    print(stack)

File: problem_solving_with_DSA/stack/C/stringRev.c
#include "stack.h"

/**
 * stringrev - reverse a string using stack 
 * 
 * @str: String to reverse
 * Return: Reverse string 
 */
char* stringrev(char *str)
{
    int len;
    
    for (len = 0; str[len] != '\0'; len++);
    
    char *temp = malloc(sizeof(char) * len+1);
    if (temp == NULL) return NULL;
    
    for (int i = 0; i < len; i++)
        push(str[i]);
    
    int i = 0;
    while(!isStackEmpty())
        temp[i++] = pop();
    
    return (temp);
}

/**
 * stringrev2 - reverse a string using stack 
 * 
 * @str: String to reverse
 * Return: Reverse string 
 */
char* stringrev2(char str[])
{
    for (int i = 0; str[i]; i++)
        push(str[i]);
    
    int i = 0;
    while(!isStackEmpty())
        str[i++] = pop();
    
    return (str);
}


int main(void)
{
    char *str = "Beloved";
    
    printf("Before => %s\n", str);
    
    char *res = stringrev(str);
    printf("After  => %s\n", res);
    
    char str2[] = "Apple"; 
    stringrev2(str2);
    printf("After  => %s\n", str2);
    
    return (0);
}